### PR TITLE
Allow an external endpoint to produce bot messages

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- moves from the template based generation of bot responses to a more general approach

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by imsurinder90** [comment]
@tmbo it would be easier to understand the changes if you add comments in newly created file and explain the changes in detail :)

**Comment by imsurinder90** [comment]
@tmbo thanks for adding comments in the new files and also migration guide which makes it quite easy to understand the code.

I want to understand in which use-case I can add endpoint and add to the interpreter. If there is a feature request created for this, can you please provide link to that so I can understand when such implementation is required.

File: rasa_core/run.py [hunk 1]
```diff
@@ -137,15 +144,51 @@ def create_input_channel(channel, port, credentials_file):
             raise Exception("Unknown input channel for running main.")
 
 
+def interpreter_from_args(nlu_model, nlu_endpoint):
```
Comment (ricwo) [inline rasa_core/run.py]: A short description and typing would be helpful

File: rasa_core/nlg/callback.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import logging
+
+from typing import Text, Any, Dict
+
+from rasa_core.nlg.generator import NaturalLanguageGenerator
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa_core/nlg/callback.py]: this is inside the package so it is better to use the fullpath here, but I changed it for the other ones :+1:

File: rasa_core/nlg/callback.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        "type": "object",
+        "properties": {
+            "text": {"type": "string"},
+            "buttons": {
+                "type": "array",
+                "items": {"type": "object"}
+            },
+            "elements": {
+                "type": "array",
+                "items": {"type": "object",}
```
Comment (ricwo) [inline rasa_core/nlg/callback.py]: @tmbo just noticed this typo

Comment (tmbo) [reply rasa_core/nlg/callback.py]: :+1:

File: rasa_core/remote.py [hunk 4]
```diff
@@ -80,43 +81,42 @@ def tracker_json(self,
                      ):
         """Retrieve a tracker's json representation from remote instance."""
 
-        url = ("{}/conversations/{}/tracker?token={}"
-               "&ignore_restarts={}"
-               "&events={}").format(self.host, sender_id, self.token,
-                                    use_history, include_events)
+        url = "/conversations/{}/tracker?&ignore_restarts={}&events={}".format(
```
Comment (ricwo) [inline rasa_core/remote.py]: the token got lost here

**Pull request merged**